research-app/src/App.vue

```diff
@@ -1213,6 +1361,369 @@ export default class App extends WWTAwareComponent {
+    // (this is trivially true are there aren't any of that type)
+    const prerequisitesMet: (msgType: string) => boolean = (msgType) => {
+      const prereqs = prerequisites[msgType];
+      if (prereqs === undefined) {
+        return true;
+      }
+      return prereqs.every(t => !messageTypes.includes(t) || finishedMessageTypes.includes(t));
+    }
+
+    // We define "root" messages as messages that either
+    // - Don't depend on another message type
+    // - Might depend on another message type, but there aren't
+    //    any of that type present.
+    //    For instance, if a user has loaded an imageset via WTML and has 
+    //    that as a background image, we need to wait until the WTML has loaded.
+    //    But they may also just have a built-in image as the background
+    const isRoot = (t: string) => {
+      const prereqs = prerequisites[t];
+      return (prereqs === undefined) || prereqs.length === 0 || prereqs.every(x => !messageTypes.includes(x));
+    }
+    const rootTypes = messageTypes.filter(isRoot);
+    
+
+    const addMessagesOfType: (msgType: string) => void = (msgType) => {
+      const messagesOfType = messages.filter(msg => getType(msg) === msgType);
+      this.messageQueue = this.messageQueue.concat(messagesOfType);
+
+      const nextTypes = messagesAdjList[msgType];
+      const completionInfo = completion[msgType];
+      if (!nextTypes) {
+        return;
+      }
+
+      const nextTypesPresent = nextTypes.filter(t => messageTypes.indexOf(t) >= 0);
+
+      if (completionInfo) {
+        const [completedType, completedMatcher] = completionInfo;
+        const handler: (msg: any) => boolean = (msg) => {
+
+          //@ts-ignore
+          const matchingMessages = messagesOfType.filter(x => completedMatcher(x, msg));
+          matchingMessages.forEach(m => {
+            messagesOfType.splice(messagesOfType.indexOf(m), 1);
+          });
+          if (messagesOfType.length === 0) {
+            this.messageHandlers.delete(msgType);
+            finishedMessageTypes.push(msgType);
+            nextTypesPresent.filter(prerequisitesMet).forEach(addMessagesOfType);
+          }
+          return true;
+        };
+        this.messageHandlers.set(completedType, handler);
+      } else {
+        finishedMessageTypes.push(msgType);
+        nextTypesPresent.filter(prerequisitesMet).forEach(addMessagesOfType);
+      }
+    }
+
+    rootTypes.forEach(addMessagesOfType);
+  }
+
+  messageQueue: Message[] = [];
+
+  @Watch("messageQueue")
+  handleMessageQueueUpdate(queue: Message[]): void {
+    while (queue.length > 0) {
+      const message = queue.shift();
+      this.onMessage(message);
+    }
+  }
+
+  stateAsUrl(): string {
+
+    const coordinatesMessage: classicPywwt.CenterOnCoordinatesMessage = {
+      event: "center_on_coordinates",
+      ra: this.wwtRARad * R2D,
+      dec: this.wwtDecRad * R2D,
+      fov: this.wwtZoomDeg / 6,
+      instant: true,
```

Comment:
This piece is failing to preserve the rotation of the view. You can demonstrate this by holding down Control and click-dragging to rotate the camera along the "roll" axis.

Reply:
Is there currently a way to access the "roll" value of the view in the app?

If not, it seems like it would be pretty straightforward to add - I'm guessing it's related to one, or both, or `angle` and `rotation` in the engine's `CameraParameters`.